expression/builtin_regexp_util.go

```diff
@@ -0,0 +1,251 @@
+	var pa regexpParam
+	var err error
+
+	pa.isNull = isNull
+	if pa.isNull {
+		pa.defaultIntVal = defaultIntVal
+		return &pa, nil
+	}
+
+	pa.col, err = bf.bufAllocator.get()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get values from input
+	if err := bf.args[id].VecEvalInt(bf.ctx, input, pa.col); err != nil {
+		return nil, err
+	}
+
+	// Check if this is a const value
+	pa.isConst = bf.args[id].ConstItem(bf.ctx.GetSessionVars().StmtCtx)
+	if pa.isConst {
+		// Initialize the const
+		pa.constIntVal, err = getColumnConstValInt(pa.col, input.NumRows())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &pa, nil
+}
+
+func getColumnConstValString(col *chunk.Column, n int) (string, error) {
+	// Precondition: col is generated from a constant expression
+	for i := 0; i < n; i++ {
+		if col.IsNull(i) {
+			continue
+		}
+		return col.GetString(i), nil
+	}
+	return "", errors.New("Can't get const string")
+}
+
+func getColumnConstValInt(col *chunk.Column, n int) (int64, error) {
+	// Precondition: col is generated from a constant expression
+	for i := 0; i < n; i++ {
+		if col.IsNull(i) {
+			continue
+		}
+		return col.GetInt64(i), nil
+	}
+	return 0, errors.New("Can't get const int")
+}
+
+// memorized regexp means the constant pattern.
+// Sometimes user may input a constant pattern, and it's unnecessary to compile
+// the regexp.Regexp each time.
+type regexpMemorizedSig struct {
+	memorizedRegexp *regexp.Regexp
+	memorizedErr    error
+}
+
+func (reg *regexpMemorizedSig) isMemorizedRegexpInitialized() bool {
+	return !(reg.memorizedRegexp == nil && reg.memorizedErr == nil)
+}
+
+func (reg *regexpMemorizedSig) initMemoizedRegexp(compile func(string) (*regexp.Regexp, error), patterns *chunk.Column, n int) {
+	// Precondition: patterns is generated from a constant expression
+	if n == 0 {
+		// If the input rownum is zero, the Regexp error shouldn't be generated.
+		return
+	}
+	for i := 0; i < n; i++ {
+		if patterns.IsNull(i) {
+			continue
+		}
+		// Compile this constant pattern, so that we can avoid this repeatable work
+		re, err := compile(patterns.GetString(i))
+		reg.memorizedRegexp = re
+		reg.memorizedErr = err
+		break
+	}
+	if !reg.isMemorizedRegexpInitialized() {
+		reg.memorizedErr = errors.New("No valid regexp pattern found")
+	}
+	if reg.memorizedErr != nil {
+		reg.memorizedRegexp = nil
+	}
+}
+
+func releaseBuffers(bf *baseBuiltinFunc, params []*regexpParam) {
+	for _, pa := range params {
+		if pa.col != nil {
+			bf.bufAllocator.put(pa.col)
+		}
+	}
+}
+
+func getBuffers(params []*regexpParam) []*chunk.Column {
+	buffers := make([]*chunk.Column, 0, 5)
+	for _, pa := range params {
+		if pa.col != nil {
+			buffers = append(buffers, pa.col)
+		}
+	}
+	return buffers
+}
+
+func isResultNull(columns []*chunk.Column, i int) bool {
+	for _, col := range columns {
+		if col.IsNull(i) {
+			return true
+		}
+	}
+	return false
+}
+
+func utf8Len(b byte) int {
+	flag := uint8(128)
```

Comment:
I doubt that TiDB already has similar util functions, if TiDB does not similar util functions, it should be defined in a more common file instead of just put it in `builtin_regexp_util.go`

Reply:
> I doubt that TiDB already has similar util functions, if TiDB does not similar util functions, it should be defined in a more common file instead of just put it in `builtin_regexp_util.go`

okok